src/v/cluster/client_quota_store.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "client_quota_serde.h"
+#include "container/fragmented_vector.h"
+
+#include <absl/algorithm/container.h>
+#include <absl/container/node_hash_map.h>
+
+namespace cluster::client_quota {
+
+class store final {
+public:
+    using container_type = absl::node_hash_map<entity_key, entity_value>;
+    using range_containter_type
+      = fragmented_vector<std::pair<entity_key, entity_value>>;
+
+    /// Upserts the given quota at the given entity key
+    /// All quota types are overwritten with the given entity_value, so on alter
+    /// operations we need to read the current state of the quota and merge it
+    /// with the alterations
+    void set_quota(const entity_key&, const std::optional<entity_value>&);
+
+    /// Returns the configured quotas for the given entity key if it exists
+    std::optional<entity_value> get_quota(const entity_key&) const;
+
+    /// Returns a list of quotas that match the given predicate
+    range_containter_type
+    range(std::function<bool(container_type::const_iterator)>&& pred) const;
+
+    /// Returns the number of quotas stored in the store
+    container_type::size_type size() const;
+
+    /// Removes all client quotas in the store
+    void clear();
+
+    /// Returns a copy of all the client quotas in the store
+    container_type all_quotas() const;
```

Comment:
Yeah, this will be used for snapshots where I'll want a copy. But I guess your argument from the other quotas PR applies here too that I could return a `const T&` here and the copying can happen caller-side.